pkg/schemadsl/parser/parser.go

```diff
@@ -67,8 +71,164 @@ Loop:
 	return rootNode
 }
 
+// consumeCaveat attempts to consume a single caveat definition.
+// ```caveat somecaveat(param1 type, param2 type) { ... }```
+func (p *sourceParser) consumeCaveat() AstNode {
+	defNode := p.startNode(dslshape.NodeTypeCaveatDefinition)
+	defer p.finishNode()
+
+	// caveat ...
+	p.consumeKeyword("caveat")
+	caveatName, ok := p.consumeTypePath()
+	if !ok {
+		return defNode
+	}
+
+	defNode.Decorate(dslshape.NodeCaveatDefinitionPredicateName, caveatName)
+
+	// Parameters:
+	// (
+	_, ok = p.consume(lexer.TokenTypeLeftParen)
+	if !ok {
+		return defNode
+	}
+
+	for {
+		paramNode, ok := p.consumeCaveatParameter()
+		if !ok {
+			return defNode
+		}
+
+		defNode.Connect(dslshape.NodeCaveatDefinitionPredicateParameters, paramNode)
+		if _, ok := p.tryConsume(lexer.TokenTypeComma); !ok {
+			break
+		}
+	}
+
+	// )
+	_, ok = p.consume(lexer.TokenTypeRightParen)
+	if !ok {
+		return defNode
+	}
+
+	// {
+	_, ok = p.consume(lexer.TokenTypeLeftBrace)
+	if !ok {
+		return defNode
+	}
+
+	exprNode, ok := p.consumeCaveatExpression()
+	if !ok {
+		return defNode
+	}
+
+	defNode.Connect(dslshape.NodeCaveatDefinitionPredicateExpession, exprNode)
+
+	// }
+	_, ok = p.consume(lexer.TokenTypeRightBrace)
+	if !ok {
+		return defNode
+	}
+
+	return defNode
+}
+
+func (p *sourceParser) consumeCaveatExpression() (AstNode, bool) {
+	exprNode := p.startNode(dslshape.NodeTypeCaveatExpession)
+	defer p.finishNode()
+
+	// Special Logic Note: Since CEL is its own language, we consume here until we have a matching
+	// close brace, and then pass ALL the found tokens to CEL's own parser to attach the expression
```

Comment:
Maybe I'm missing something but we don't seem to use any of the CEL language tokens here?

Reply:
See the comment: we just consume until we see the proper braces and pass the whole expression to CEL. We need to have the tokens understood by the lexer so we can do this, but we don't *use* them.